graphql/discovery-query.graphql

```diff
@@ -0,0 +1,66 @@
+fragment TargetFields on Node {
+    name
+    nodeType
+    ... on TargetNode {
+        target {
+            alias
+            serviceUri
+            labels
+            annotations {
+                cryostat
+                platform
+            }
+        }
+    }
+}
+
+# Arbitrarily query up to 8 levels deep. GraphQL doesn't support arbitrary recursive
+# querying for nested/recurisve structures - the query has to have an explicit depth
```

Comment:
I think I remember you mentioning this being an issue, could you explain why again?

Reply:
I can define a structure in GraphQL that looks something like (pseudo"code"):

```
A {
  B[] b
  String name
}

B {
  C[] c
  String name
}

C {
  String name
}
```

We could use this to do something like:

```
var c = C("leaf 1")
var c = C("leaf 1")
var c = C("leaf 1")

Reply:
Okay, that makes sense. Would you say this is the biggest potential issue (as far as we know) if we do decide to use GraphQL for batch operations? From first impressions, I like the JSON-esque queries and it looks like mutations have us covered when it comes to performing actions on Cryostat. The `DataFetchers` themselves don't seem overly complex either.

Reply:
Yes, this seems to be the biggest limitation IMO. Luckily it only really matters for the querying side of things - mutations can operate on arbitrarily nested structures just fine, the only problem is that we can't return results with arbitrarily nested structures. For responses to mutations that probably doesn't matter, but for querying what exists it's definitely a limitation.

In practise we could probably just allow depths up to say 16 or 32 or some arbitrary "large" depth and it should cover us in the practical use cases we support, and we can document that limitation.